File: 2020/julia/src/Day6.jl
module Day6
include("Aoc.jl")
using .Aoc

groups = split(strip(Aoc.input_string(6)), "\n\n")
part1() = mapreduce(g -> replace(g, "\n" => "")|>unique|>length, +, groups)
part2() = mapreduce(group -> length(reduce(intersect, [Set(person) for person in split(group, "\n")])), +, groups)
end

File: 2020/typescript/src/days/day6.ts
import {Day, sum} from "../aoc";

export class Day6 extends Day {
    private groups: string[] = [];

    day = (): number => 6;

    part1 = (): string => this.groups
        .map(group => group.replaceAll(/\n/g, '').split(''))
        .map(group => new Set(group).size)
        .reduce(sum)
        .toString()

    part2 = (): string => this.groups
        .map(group => group
            .split(/\r?\n/)
            .map(person => new Set(person.split('')))
            .reduce((personA, personB) => new Set([...personA].filter(ans => personB.has(ans))))
            .size
        )
        .reduce(sum)
        .toString()

    setup = () => this.groups = this.input.join("\n").trim().split(/\r?\n\r?\n/);
}

